### PR TITLE
Add new parser script to parse Unified Campaigns

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes 

#### Status
ready

#### Description
This PR provides a new script which can be used to parse Unified campaigns either from given input json file or from Unified MongoDB. Here is a typical usage:
```
bin/adhoc-scripts/parseUnifiedCampaigns.py --help
usage: PROG [-h] [--fin FIN] [--verbose VERBOSE] [--url URL] [--dburi DBURI]
            [--dbname DBNAME] [--dbcoll DBCOLL]

optional arguments:
  -h, --help         show this help message and exit
  --fin FIN          Input file
  --verbose VERBOSE  verbosity level
  --url URL          url of server where we should upload campaign configs
  --dburi DBURI      MongoDB URI
  --dbname DBNAME    MongoDB database
  --dbcoll DBCOLL    MongoDB collection

# parse unified campaigns and stored them to external file
./parseUnifiedCampaigns.py --dburi=mongodb://vocms0274.cern.ch:27017 --dbname=unified --dbcoll=campaignsConfiguration > ~/public/UnifiedCampaigns.txt

# parse unified campaigns and push results into couchdb
./parseUnifiedCampaigns.py --dburi=mongodb://vocms0274.cern.ch:27017 --dbname=unified --dbcoll=campaignsConfiguration --url=https://vk-reqmgr.cern.ch/reqmgr2/data

```

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
#9139 

#### External dependencies / deployment changes
No

### CODE REVIEW

**Comment by amaltaro** [comment]
@vkuznet can you please post here one example of WMCore campaign configuration?

Then @sharad1126 and @thongonary can have a look and see whether it suits all the current Ops needs.

**Comment by vkuznet** [comment]
Here is a file which you can look for campaign configuration:
/afs/cern.ch/user/v/valya/public/UnifiedCampaigns.txt

It is what we propose to push to CouchDB, and it follows the proposed schema: https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154

Do you need to see campaign configuration which are stored in Unified DB?

We discussed this here: https://github.com/dmwm/WMCore/issues/9139

**Comment by amaltaro** [comment]
> It is what we propose to push to CouchDB, and it follows the proposed schema: https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154

This is the internal data structure schema for the workflow data placement monitoring, not the campaign schema.

Valentin, let's follow up the campaign schema discussion here: https://github.com/dmwm/WMCore/issues/9139
Once we are happy with something, we can then get back to P&R and discuss it with then.

**Comment by vkuznet** [comment]
ok, I commented over there.

**Comment by amaltaro** [comment]
@vkuznet can you please update your script such that it returns a campaign formatted as discussed in this thread:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/FwkJobReport/Report.py#L1044

In addition to that, I'd suggest to use upper camel case convention for those parameters. So the final WMCore campaign configuration should look like:
```
{
 "CampaignName": "blah2019",
 "PrimaryAAA": boolean, # default false
 "SecondaryAAA": boolean, # default false
 "SiteWhiteList": ["T1_*", "T2_*"],
 "SiteBlackList": [...],
 "SecondaryLocation": ["list with sites to hold a whole copy of the dataset"],
 "Secondaries": {"datasetA": ["list of sites for SiteWhitelist"], "datasetB": ["list of sites for SiteWhitelist"]},
 "MaxCopies": integer
}
```

Please make the final list of campaigns available such that we can do some QA before proceeding to the next step.

**Comment by vkuznet** [comment]
Alan, few things:
- I think you meant that discussion about campaign is happening in this thread #9139 , right?
- we need some logic how to parse and extract requested attributes, I'll post a question on #9139

**Comment by vkuznet** [comment]
Alan,
I adjusted the code of the parser and stored records from Unified MongoDB over here:
/afs/cern.ch/user/v/valya/public/UnifiedCampaigns.txt

Please have a look and let me know. I don't mind to send it to testbed or your VM, but it would be nice to verify records before sending them to given URL (in case of your VM please remind me its name).

**Comment by amaltaro** [comment]
Valentin,
having a glimpse at your text file, I could spot 4 issues:
1. many campaign duplication (e.g.: RunIIFall17pLHE)
2. two keys defining campaign name (please keep only `CampaignName`, as proposed in the schema)
3. `Secondaries` does not follow the schema proposed, I found empty dictionaries instead of empty list (e.g.: PhaseIITDRFall17DR)
4. Again `Secondaries` doesn't match the schema when there is a `SecondaryLocation`, your campaign is:
```
{"Campaign": "RunIISpring16DR80", "MaxCopies": 2, "SiteWhiteList": ["T1_US_FNAL", "T2_US_Purdue", "T2_US_Wisconsin", "T2_US_UCSD", "T2_US_Vanderbilt", "T2_US_Florida", "T2_US_Caltech", "T2_US_Nebraska", "T2_US_MIT"], "SecondaryLocation": [], "CampaignName": null, "SecondaryAAA": false, "SiteBlackList": [], "Secondaries": {"/Neutrino_E-10_gun/RunIISpring15PrePremix-PU2016_80X_mcRun2_asymptotic_v14-v2/GEN-SIM-DIGI-RAW": {"SecondaryLocation": ["T1_US_FNAL"], "secondary_AAA": true}}, "PrimaryAAA": false}
```
when the expected result is:
```
{"MaxCopies": 2, "SiteWhiteList": ["T1_US_FNAL", "T2_US_Purdue", "T2_US_Wisconsin", "T2_US_UCSD", "T2_US_Vanderbilt", "T2_US_Florida", "T2_US_Caltech", "T2_US_Nebraska", "T2_US_MIT"], "SecondaryLocation": ["T1_US_FNAL"], "CampaignName": "RunIISpring16DR80", "SecondaryAAA": true, "SiteBlackList": [], "Secondaries": {"/Neutrino_E-10_gun/RunIISpring15PrePremix-PU2016_80X_mcRun2_asymptotic_v14-v2/GEN-SIM-DIGI-RAW": []}, "PrimaryAAA": false}
```

These are issues that I could spot. I also see we didn't account for having `secondary_AAA` inside the secondaries key. I'll update the description and how we have to parse the original campaign.json doc.

FYI: Sharad is trying to run some cleanup on the campaign.json file and it would also be slightly more homogeneous, he has this PR pending: https://github.com/CMSCompOps/WmAgentScripts/pull/430

**Comment by amaltaro** [comment]
Valentin, you might want to check the first comment on the GH issue #9139 ; I updated it with the proposed schema and what data type; which key gets mapped to it, and further logic in case it's found in multiple times under the same campaign configuration. 

Of course, all that according to what was discussed with Ops and as far as my (and theirs) understanding go...

File: bin/adhoc-scripts/parseUnifiedCampaigns.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            "secondary_AAA": "SecondaryAAA",
+            "primary_AAA": "PrimaryAAA",
+            "maxcopies": "MaxCopies",
+            "secondaries": "Secondaries",
+            }
+    # campaign schema dict
+    confRec = {'SiteWhiteList': [], 'SiteBlackList': [],
+            'CampaignName': None, 'PrimaryAAA': False,
+            'SecondaryAAA': False, 'SecondaryLocation': [],
+            'Secondaries': {}, 'MaxCopies': 0}
```
Comment (amaltaro) [inline bin/adhoc-scripts/parseUnifiedCampaigns.py]: `MaxCopies` default value is 1

File: bin/adhoc-scripts/parseUnifiedCampaigns.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                        if kkk.lower() == 'secondaryaaa' or kkk.lower() == 'secondary_aaa':
+                            secAAA = True if vvv else secAAA
+            if verbose > 2:
+                print("conf: %s" % conf)
+            sec = dict(conf['Secondaries'])
+            sdict = {}
+            secLoc = []
+            for key, val in sec.items():
+                wlist = val.get('SiteWhitelist', [])
+                blist = val.get('BlackWhitelist', [])
```
Comment (amaltaro) [inline bin/adhoc-scripts/parseUnifiedCampaigns.py]: `SiteBlacklist`

**Review by amaltaro — CHANGES_REQUESTED** [review]
I spotted 2 more issues, as highlighted in the code.

Then there is another case - which I didn't spot in the code - where the final campaign configuration is wrong for `RunIISpring16DR80`. According to the schema definition and where values have to be taken from:
* `SecondaryAAA` should be `True`, since there is a `secondary_AAA: true` under the `secondaries key.
* and the `Secondaries` pileup dataset should NOT have an empty list location, but instead it should be `T1_US_FNAL`, as per the inner `SecondaryLocation` key/value.

In short, MaxCopies, SiteBlackList, SecondaryAAA and Secondaries have problems.

**Comment by amaltaro** [comment]
@vkuznet in addition to the comments above, please reformat this code and get rid of those pylint/pep8 issues reported. It's a new file, so we better start with the right foot.

**Comment by vkuznet** [comment]
test this please

**Comment by vkuznet** [comment]
Alan, code is fixed and passed pylint/pycodestyle (except one complain about script name, I don't know how you'd like to name scripts though). And, I updated /afs/cern.ch/user/v/valya/public/UnifiedCampaigns.txt area with new context. Please check and review.

File: bin/adhoc-scripts/parseUnifiedCampaigns.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+                    secAAA = True if val else secAAA
+                if key == 'parameters':
+                    for kkk, vvv in crec[key].items():
+                        if kkk in conf.keys():
+                            conf[kkk] = vvv
+                        if kkk in remap.keys():
+                            conf[remap[kkk]] = vvv
+                        if kkk.lower() == 'sitewhitelist':
+                            whiteList = intersect(whiteList, vvv)
+                        if kkk.lower() == 'blackwhitelist':
```
Comment (amaltaro) [inline bin/adhoc-scripts/parseUnifiedCampaigns.py]: `SiteBlacklist`

**Review by amaltaro — CHANGES_REQUESTED** [review]

File: bin/adhoc-scripts/parseUnifiedCampaigns.py (hunk 1 above)
Comment (vkuznet) [reply bin/adhoc-scripts/parseUnifiedCampaigns.py]: fixed

File: bin/adhoc-scripts/parseUnifiedCampaigns.py (hunk 2 above)
Comment (vkuznet) [reply bin/adhoc-scripts/parseUnifiedCampaigns.py]: fixed

**Comment by amaltaro** [comment]
Please produce a brand new list of campaigns with this up-to-date script.

**Comment by vkuznet** [comment]
tried but unified node is not accessible now, seems like under VM reboot
procedure.

**Comment by amaltaro** [comment]
It looks like you have just updated the AFS text file(?) Please confirm and I will have another look at it.

**Comment by vkuznet** [comment]
Yes, I was about to write about it. Please check it.

**Comment by amaltaro** [comment]
@vkuznet in order to be efficient and move forward, I made a new fix here: #9306
I'm closing this one.
